### DISCUSSION TITLE
Warnings: A descriptor exists for fd=..., but it is not a legacy descriptor. Returning NULL.

### DISCUSSION THREAD

**Discussion by cdoepmann** [shadow/shadow#2071] [Q&A] [answered]
I'm currently using tornettools + shadow to carry out simulations of the Tor network. When doing so, I am getting tons of warnings of the following kind:

```
30:37:45.545683 [32623:shadow-worker] 00:48:12.288925012 [WARN] [server29:11.0.10.235] [process.rs:304] [shadow_rs::host::process::export] A descriptor exists for fd=1149, but it is not a legacy descriptor. Returning NULL.
30:37:45.939976 [30027:shadow-worker] 00:48:12.294786011 [WARN] [server174:11.0.10.158] [process.rs:304] [shadow_rs::host::process::export] A descriptor exists for fd=7402, but it is not a legacy descriptor. Returning NULL.
30:37:46.032076 [30073:shadow-worker] 00:48:12.296260017 [WARN] [server20:11.0.10.187] [process.rs:304] [shadow_rs::host::process::export] A descriptor exists for fd=4645, but it is not a legacy descriptor. Returning NULL.
30:37:47.310421 [28896:shadow-worker] 00:48:12.317076024 [WARN] [server158:11.0.10.140] [process.rs:304] [shadow_rs::host::process::export] A descriptor exists for fd=6867, but it is not a legacy descriptor. Returning NULL.
30:37:47.698113 [30764:shadow-worker] 00:48:12.326351023 [WARN] [server233:11.0.10.224] [process.rs:304] [shadow_rs::host::process::export] A descriptor exists for fd=2825, but it is not a legacy descriptor. Returning NULL.
30:37:48.174519 [32540:shadow-worker] 00:48:12.338859004 [WARN] [server124:11.0.10.103] [process.rs:304] [shadow_rs::host::process::export] A descriptor exists for fd=2842, but it is not a legacy descriptor. Returning NULL.
30:37:48.581346 [28683:shadow-worker] 00:48:12.349298003 [WARN] [server88:11.0.11.46] [process.rs:304] [shadow_rs::host::process::export] A descriptor exists for fd=1686, but it is not a legacy descriptor. Returning NULL.
30:37:48.923909 [30270:shadow-worker] 00:48:12.359411046 [WARN] [server221:11.0.10.211] [process.rs:304] [shadow_rs::host::process::export] A descriptor exists for fd=4554, but it is not a legacy descriptor. Returning NULL.
30:37:49.263234 [32616:shadow-worker] 00:48:12.369916026 [WARN] [server138:11.0.10.118] [process.rs:304] [shadow_rs::host::process::export] A descriptor exists for fd=6543, but it is not a legacy descriptor. Returning NULL.
```

These all happen on the "server" nodes (tgen servers). I wonder what might be the cause. Also, I am not sure what implications this might have for the simulation results. At first sight, it doesn't seem to break everything, but I'm not exactly sure. Is there any experience on that?

**Comment by stevenengler** [comment, marked as answer]
In Tor simulations these warnings are probably all caused by Unix sockets during glibc's address resolution, and in this case are not an issue. I'll make a PR to fix that warning today, and then if you still see these warnings after that fix, then please make an issue.

The actual cause of the warning in this case is that our Unix socket type is written in Rust, but our `connect()` syscall handler implementation is written in C. Since the C `connect()` syscall handler doesn't know what to do with a Rust socket type, it logs that warning and returns `EBADF`. But for address resolution, the `connect()` syscall for the unix socket is expected to return an error anyways since we aren't running any DNS-related daemons in Shadow.

**Reply by cdoepmann** [reply]
Great, thank you for the clarification and the thorough explanation!

**Reply by stevenengler** [reply]
We decided for now in #2073 to change the warning message specifically for syscalls using unix sockets until we finish implementing them. So now it should show warning messages such as `"connect() syscall not yet supported for fd X of type Unix; Returning ENOSYS"` instead. If you still see the `"A descriptor exists for fd=X, but it is not a legacy descriptor. Returning NULL"` messages, please make an issue for it.
